### PR TITLE
test(RHTAPWATCH-1067): Unit tests for getUserNamespaces

### PR DESCRIPTION
**Pull request by kousalyalakshmanan**
Add unit test for func getUserNamespaces

### CODE REVIEW

File: cmd/main.go [hunk 1]
```diff
@@ -151,10 +151,11 @@ func getUserNamespaces(e *echo.Echo, nameReq labels.Requirement) ([]core.Namespa
 	if err != nil {
 		e.Logger.Fatal(err)
 	}
-	req, _ := labels.NewRequirement("konflux.ci/type", selection.In, []string{"user"})
-	selector := labels.NewSelector().Add(*req)
-	selector.Add(nameReq)
+	// create a label selector based on the requirement
```
Comment (Omeramsc) [inline cmd/main.go]: Why are we changing an already working function?

File: cmd/main_test.go [hunk 2]
```diff
@@ -301,3 +304,60 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+		// Check if the actual namespaces match the expected namespaces
+		Expect(actualNamespaces).To(ConsistOf(expectedNamespaces))
+	},
+	Entry(
+		"Get all user namespaces",
+		"konflux.ci/type",
+		[]string{"user"},
+		[]string{"test-tenant", "test-tenant-2", "test-tenant-3"},
+	),
+	// This test checks if we can retrieve all namespaces labeled as `user` namespaces.
```
Comment (Omeramsc) [inline cmd/main_test.go]: This text should mostly be summarized in the Entry's name IMO.

File: cmd/main_test.go [hunk 3]
```diff
@@ -301,3 +304,60 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+	// Expected is it should return only the "test-tenant" namespace.
+
+	Entry(
+		"Get multiple specific user namespaces",
+		"kubernetes.io/metadata.name",
+		[]string{"test-tenant", "test-tenant-2"},
+		[]string{"test-tenant", "test-tenant-2"},
+	),
+	// This test checks if we can retrieve multiple specific namespaces by their names given.
+	// Exepected is it should return both "test-tenant" and "test-tenant-2" namespaces.
```
Comment (Omeramsc) [inline cmd/main_test.go]: It should be understandable from the `expectedNamespaces` part of the Entry, no need to mention it.

File: cmd/main_test.go [hunk 4]
```diff
@@ -301,3 +304,60 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+		"kubernetes.io/metadata.name",
+		[]string{"test-tenant"},
+		[]string{"test-tenant"},
+	),
+	// This test checks if we can retrieve a single specific namespace by its name given.
+	// Expected is it should return only the "test-tenant" namespace.
+
+	Entry(
+		"Get multiple specific user namespaces",
+		"kubernetes.io/metadata.name",
```
Comment (Omeramsc) [inline cmd/main_test.go]: The function checks for `konflux.ci/type` + other labels (as input)

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Overall good progress, but I am not sure why a change in the  function itself is required

**Comment by kousalyalakshmanan** [comment]
> Overall good progress, but I am not sure why a change in the function itself is required

I saw your comments on the PR thanks. The reason why I had to modify the main function was: take for eg to get specific user namespace test case- the test is expecting only one namespace (“test-tenant”) but is returning all three namespaces (test-tenant, test-tenant2, test-tenant3). I felt that the label selector is not working as expected and thats why I added filteredNamespaces func in the main.go.

**Comment by kousalyalakshmanan** [comment]
@Omeramsc I tried the method that you suggested

`I think you get all the namespaces because all of them have those labels. I
t makes sense since we created them to be used in the functional tests, if they would not have those labels they won't pass.It means we don't need to change the function, as it works as it should. what we need to change is the namespaces we create. or rather - create a new namespace (either for this test specifically or for the whole beforesuite, see what works better), that doesn't have this label. then you can see it is not returned.`

**Comment by kousalyalakshmanan** [comment]
> @Omeramsc I tried the method that you suggested
> 
> `I think you get all the namespaces because all of them have those labels. I t makes sense since we created them to be used in the functional tests, if they would not have those labels they won't pass.It means we don't need to change the function, as it works as it should. what we need to change is the namespaces we create. or rather - create a new namespace (either for this test specifically or for the whole beforesuite, see what works better), that doesn't have this label. then you can see it is not returned.`

PTAL again. test cases pass.

File: cmd/main.go (hunk 1 above)
Comment (kousalyalakshmanan) [reply cmd/main.go]: reverted

File: cmd/main_test.go (hunk 2 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: updated

**Comment by hmariset** [comment]
lgtm

File: cmd/main_test.go [hunk 5]
```diff
@@ -251,6 +264,7 @@ var _ = BeforeSuite(func() {
 	createNamespace(k8sClient, "test-tenant")
 	createNamespace(k8sClient, "test-tenant-2")
 	createNamespace(k8sClient, "test-tenant-3")
+	createNamespaceWithoutLabels(k8sClient, "test-tenant-no-labels")
```
Comment (Omeramsc) [inline cmd/main_test.go]: Since related only to the unit-tests, we might want to use `beforeeach`  and `aftereach` to create *and* remove this namespace

File: cmd/main_test.go [hunk 6]
```diff
@@ -301,3 +315,59 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+		[]string{"test-tenant"},
+	),
+	Entry(
+		"Get multiple specific user namespaces",
+		"kubernetes.io/metadata.name",
+		[]string{"test-tenant", "test-tenant-2"},
+		[]string{"test-tenant", "test-tenant-2"},
+	),
+	Entry(
+		"Get non-existent namespace/ checks to retrieve a namespace that doesn't exist",
```
Comment (Omeramsc) [inline cmd/main_test.go]: This is a bit confusing, as we don't get or retrieve anything here, we are just attempting to.
 so I'd rephrase it to be something along the lines of "returns an empty string when the label mentions a namespace that does not exist"

File: cmd/main_test.go [hunk 7]
```diff
@@ -301,3 +315,59 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+			req, err = labels.NewRequirement(labelKey, selection.Exists, []string{})
+		}
+		Expect(err).NotTo(HaveOccurred(), "Error creating label requirement")
+
+		namespaces, err := getUserNamespaces(e, *req)
+		Expect(err).NotTo(HaveOccurred(), "Error getting user namespaces")
+
+		var actualNamespaces []string
+		for _, ns := range namespaces {
+			actualNamespaces = append(actualNamespaces, ns.Name)
```
Comment (Omeramsc) [inline cmd/main_test.go]: I think the entries will be more readable if we use an actual namespace object. it will also make the test more accurate, actually testing against the whole object returned. 
Take a look at Homaja's #40 , she created a fixture to help with that, we can use it here as well.

File: cmd/main_test.go [hunk 8]
```diff
@@ -301,3 +315,59 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+		[]string{"test-tenant", "test-tenant-2"},
+		[]string{"test-tenant", "test-tenant-2"},
+	),
+	Entry(
+		"Get non-existent namespace/ checks to retrieve a namespace that doesn't exist",
+		"kubernetes.io/metadata.name",
+		[]string{"non-existent-namespace"},
+		[]string{},
+	),
+)
```
Comment (Omeramsc) [inline cmd/main_test.go]: Why did we create `createNamespaceWithoutLabels` if we don't have a testcase for it?

**Review by Omeramsc — CHANGES_REQUESTED** [review]
It looks much better now; I added a few more comments

File: cmd/main_test.go (hunk 7 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: I see that Homaja has updated her code again. I will work on adding beforeeach and aftereach.

File: cmd/main_test.go [hunk 9]
```diff
@@ -301,3 +304,53 @@ var _ = DescribeTable("TestRunAccessCheck", func(user string, namespace string,
...
+	Entry(
+		"Get specific user namespace",
+		"kubernetes.io/metadata.name",
+		[]string{"test-tenant"},
+		[]string{"test-tenant"},
+	),
+	Entry(
+		"Get multiple specific user namespaces",
+		"kubernetes.io/metadata.name",
+		[]string{"test-tenant", "test-tenant-2"},
```
Comment (Omeramsc) [inline cmd/main_test.go]: are those label values? it looks like a list of namespaces

File: cmd/main_test.go (hunk 7 above)
Comment (Omeramsc) [reply cmd/main_test.go]: Please tag me for review when done, the test will change massively as a result so reviewing it now is a bit redundant

**Review by Omeramsc — CHANGES_REQUESTED** [review]

**Comment by kousalyalakshmanan** [comment]
@Omeramsc PTAL. Three test cases fail out of the 4 because the  function `getUserNamespaces` appears to be returning all user namespaces, regardless of the specific request. The same error I said in the beginning.

File: cmd/main_test.go [hunk 10]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
 	})
 })
+
+var _ = Describe("GetUserNamespaces", func() {
+	var e *echo.Echo
+	var req *labels.Requirement
+	var expectedNamespaces []string
+
+	BeforeEach(func() {
+		e = echo.New()
```
Comment (Omeramsc) [inline cmd/main_test.go]: I am not sure we need to create the echo server for each test case. it can be moved to the former section of the test and applied to all testcases.

File: cmd/main_test.go [hunk 11]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+	})
+
+	Context("When querying for a specific user namespace", func() {
+		BeforeEach(func() {
```
Comment (Omeramsc) [inline cmd/main_test.go]: The namespaces for the test should be created here.

File: cmd/main_test.go [hunk 12]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+			var err error
+			req, err = labels.NewRequirement("kubernetes.io/metadata.name", selection.In, []string{"test-tenant"})
```
Comment (Omeramsc) [inline cmd/main_test.go]: Where are we creating test-tenant?

File: cmd/main_test.go [hunk 13]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+		})
+
+		AfterEach(func() {
+			req = nil
+			expectedNamespaces = nil
+		})
+	})
+
+	Context("When querying for multiple specific user namespaces", func() {
+		BeforeEach(func() {
```
Comment (Omeramsc) [inline cmd/main_test.go]: same

File: cmd/main_test.go [hunk 14]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+
+			var actualNamespaces []string
+			for _, ns := range namespaces {
+				actualNamespaces = append(actualNamespaces, ns.Name)
+			}
+
+			Expect(actualNamespaces).To(ConsistOf(expectedNamespaces))
+		})
+
+		AfterEach(func() {
```
Comment (Omeramsc) [inline cmd/main_test.go:651]: After creating the namespaces, they should be removed in this section.

File: cmd/main_test.go [hunk 15]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+			req = nil
+			expectedNamespaces = nil
```
Comment (Omeramsc) [inline cmd/main_test.go]: no need to reset the variables.

File: cmd/main_test.go [hunk 16]
```diff
@@ -476,3 +478,116 @@ var _ = DescribeTable("GetWorkspacesWithAccess querying for workspaces with acce
...
+			Expect(namespaces).To(BeEmpty())
+		})
+
+		AfterEach(func() {
+			req = nil
+			expectedNamespaces = nil
+		})
+	})
+
+	// This test case alone passes becasue getUserNamespaces function is always returning all user namespaces regardless of the input requirement.
```
Comment (Omeramsc) [inline cmd/main_test.go]: It does not. `getUserNamespaces` return all namespaces with the `konflux.ci/type`(=user) label and any additional labels you've given it. 
if it returns all namespaces, something in the resource management for the test is wrong.

**Review by Omeramsc — CHANGES_REQUESTED** [review]
> @Omeramsc PTAL. Three test cases fail out of the 4 because the function `getUserNamespaces` appears to be returning all user namespaces, regardless of the specific request. The same error I said in the beginning.

1. The BeforeSuite is still applied to all tests. this is not desireable, I'll push a change seperating the unit tests from the functional tests.
2. Probably a result of the former point, but in all tests you never created a namespace. the namespaces from the BeforeSuite are configured to pass the functional tests, so the fact they always return is a given. you shouldn't use them, you should create your own in the beforeEach
3. the echo server however is not needed in the beforeEach
4. resetting variables is in the afterEach is not needed, you need to reset (=remove) the namespaces you create. 

If you have any more questions about why we use the beforeEach and afterEach for those tests and resource management, please don't hesitate to ask me before applying more changes to the code.

File: cmd/main_test.go (hunk 16 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: Okay will check that out

**Comment by kousalyalakshmanan** [comment]
This PR is blocked by [RHTAPWATCH-1237](https://issues.redhat.com/browse/RHTAPWATCH-1237)	Reorganize workspace-manager test suites.

File: cmd/main.go [hunk 17]
```diff
@@ -153,7 +153,7 @@ func getUserNamespaces(e *echo.Echo, nameReq labels.Requirement) ([]core.Namespa
 	}
 	req, _ := labels.NewRequirement("konflux.ci/type", selection.In, []string{"user"})
 	selector := labels.NewSelector().Add(*req)
-	selector.Add(nameReq)
+	selector = selector.Add(nameReq)
```
Comment (Omeramsc) [inline cmd/main.go:156]: What is the impact of this change? 
can you show  a log example of the difference in `selector` between those?

File: cmd/main_test.go [hunk 18]
```diff
@@ -615,3 +617,117 @@ var _ = DescribeTable("TestGetNamespacesWithAccess", func(allNamespaces []k8sapi
...
+
+		BeforeEach(func() {
+			namesToCreate := []string{"test-ns-1", "test-ns-2", "test-ns-3"}
+			for _, name := range namesToCreate {
+				ns, err := createNamespace(k8sClient, name)
+				Expect(err).NotTo(HaveOccurred(), fmt.Sprintf("Error creating the namespace %s", name))
+				createdNamespaces = append(createdNamespaces, ns.Name)
+			}
+
+			var err error
```
Comment (Omeramsc) [inline cmd/main_test.go]: can't we use `:=` in the following line?

File: cmd/main_test.go [hunk 19]
```diff
@@ -615,3 +617,117 @@ var _ = DescribeTable("TestGetNamespacesWithAccess", func(allNamespaces []k8sapi
...
+		AfterEach(func() {
+			for _, ns := range createdNamespaces {
+				deleteNamespace(k8sClient, ns)
+			}
+		})
+	})
+
+	Context("When querying for namespaces using NotIn", func() {
+
+		BeforeEach(func() {
```
Comment (Omeramsc) [inline cmd/main_test.go]: We are using BeforeEach on singular tests. I think we should either use it on multiple cases or ditch it altogether

File: cmd/main.go (hunk 17 above)
Comment (Omeramsc) [reply cmd/main.go:156]: Some extra context:
before
```
line155: Selector value: konflux.ci/type in (user)
line156: Selector value: konflux.ci/type in (user)
exclude-2)
```

after
```
line155: Selector value: konflux.ci/type in (user)
line156: Selector value: konflux.ci/type in (user),kubernetes.io/metadata.name in (in-test-1,in-test-2)
```

Comment (Omeramsc) [reply cmd/main.go:156]: It does seem to be the right way to use it, good catch!

**Review by Omeramsc — CHANGES_REQUESTED** [review]
Good catch, I'm glad to see it moving forward. added some small requests

File: cmd/main_test.go (hunk 18 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: @Omeramsc which lines exactly are you talking about?

Comment (Omeramsc) [reply cmd/main_test.go]: 638
```req, err = labels.NewRequirement("kubernetes.io/metadata.name", selection.Exists, []string{})```

File: cmd/main_test.go (hunk 19 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: I didn't get it. @Omeramsc Could you please explain again? why I should remove beforeeach?

File: cmd/main_test.go (hunk 18 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: @Omeramsc When I try to update := on line 638 it throws this error
image

File: cmd/main.go (hunk 17 above)
Comment (kousalyalakshmanan) [reply cmd/main.go:156]: Okay. Added

File: cmd/main_test.go (hunk 18 above)
Comment (Omeramsc) [reply cmd/main_test.go]: this error applies to `req`, not `err`. debug why it is showing you this error when you do reference req earlier.

File: cmd/main_test.go (hunk 19 above)
Comment (Omeramsc) [reply cmd/main_test.go]: `BeforeEach` is used when you need to perform any setup tasks that should be done before every individual test case, ensuring that tests are reproducible and independent. 
in this patch you use `BeforeEach` on singular tests, as every test has only 1 single testcase. it makes it redundant, and not much different than having the setup as-is. 
The idea behind BeforeEach is to have the setup once for multiple testcases, similar (but not exactly) to fixtures.

File: cmd/main_test.go (hunk 18 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: I think its becasue req is already defined outside the `beforeach` fun. 
:= I see is been used for declaring and initializing new variables. since we have declared req already outside beforeeach we can only use = in here.

File: cmd/main_test.go (hunk 19 above)
Comment (kousalyalakshmanan) [reply cmd/main_test.go]: @Omeramsc I have updated the code now.
Changes that I made are:
-  I removed var err and req that were defined outside of `beforeeach` and using := now for declaring and initializing new variables in a single line like you suggested earlier.
- I removed the `beforeeach` block on the testcases and moved it directly inside the `It` block
- Keeping the `Aftereach` block as it is fro cleanup.

**Review by Omeramsc — APPROVED** [review]
LGTM
